### ISSUE TITLE
Feature request: Expandable notifications (inline + toast) 

### ISSUE THREAD

**Issue by mceledonia** [patternfly/patternfly-design#892]
We've had a few use cases pop up for an expandable function within notifications. During the last alert design refactor I came up with some designs around this which I'll attach below, these might work but it's worth noting they were sort of a quick concept rather than a fully thought out design.

Given the recent product requests I feel like it would be a good idea to get this on the roadmap. 

<img width="433" alt="image" src="https://user-images.githubusercontent.com/30935938/87072712-40afea00-c1ea-11ea-9d7a-7190a0459232.png">

Some initial notes:
I'd see the expanded content choice being pretty flexible, maybe it starts with a headline and expands to show description/actions, maybe it starts with headline and 1 line of description and expands to show full description, etc. etc.

**Comment by gdoyle1** [comment]
@mceledonia I have something super similar to this although I'd think of it more of a "toast modal"... similar to when you upload files to google drive
<img width="412" alt="Screen Shot 2020-07-09 at 2 02 53 PM" src="https://user-images.githubusercontent.com/30870144/87074461-e5332b80-c1ec-11ea-9620-d745089fa58f.png">

**Comment by mceledonia** [comment]
@gdoyle1 Yeah definitely, that seems to me like an expandable toast which also has a much more flexible range of content that it can hold. Want to get @mcarrano's thoughts on this too!

**Comment by ncameronbritt** [comment]
I'd like something similar for inline alerts:
![Screen Shot 2020-07-09 at 2 44 16 PM](https://user-images.githubusercontent.com/8974593/87078511-bf108a00-c1f2-11ea-8fb0-4291203b7abd.png)

**Comment by mcarrano** [comment]
@mceledonia Sounds like there are multiple use cases for something like this. Can you also take a look at this request: https://github.com/patternfly/patternfly-design/issues/874  The use case is entirely different but it strikes me that these are pretty much the same things.

**Comment by mcarrano** [comment]
@jessiehuff @jgiardino We were discussing this request today as an enhancement to our toast alert and I wanted to run this past you to see if this would be an accessible solution for screen reader or keyboard users as I know this was a general concern with actions inside toasts.

**Comment by jgiardino** [comment]
My recommendation varies depending on whether this is a toast or not. 

If not a toast, then the [WAI-ARIA disclosure pattern](https://www.w3.org/TR/wai-aria-practices-1.2/#disclosure) could be used. 

But toasts are a bit more complicated. I would argue that interactive elements in a toast notification are not easily accessible by screen reader users or keyboard only users, assuming that toast notifications are not persistent. Depending on where the current focus is on the page and the number of elements on the page, it's very likely that these users would not be able to navigate to a toast notification in time before it expires.

Additionally, the screen reader does not consume the contents of the notification by interacting with it, but rather, the contents are announced to the user. When contents are announced to the user, the labels of interactive contents are announced but the user cannot interact with them. So outside the context of this enhancement, if there is functionality provided in a toast, then ideally there would be screen-reader-only contents that describe how to access this same functionality (e.g. "Navigate to the Buffalo Dip details page to celebrate with a toast or make more."), and maybe the interactive elements should have an aria-hidden to avoid announcing labels out of context of their controls.

Ideally, to make these toasts fully accessible, they would also be available in a drawer, so that when (any) users see them come and go, they are able to find the same message and actions in a drawer.

With all that, here are some possible ideas to make the toast version of this accessible:
- option 1 - The toggle that displays more contents has an aria-hidden, and the contents are provided via the screen-reader-only styles. This would enable them to be announced to the screen reader user while not being visible.
- option 2 -  The toggle that displays more contents has an aria-hidden, and an additional screen-reader-only message instructs the user to access the notification drawer for more details. 

The advantage of option 1 is that all contents are announced to the user without requiring them to find the notification drawer. But this can also be a disadvantage if there are a lot of contents to listen to. And another advantage of option 2 is that the user is informed about how to access notifications, which enables them to take advantage of other benefits provided in the drawer that are otherwise not accessible to them in the alert.

**Comment by mcarrano** [comment]
Thanks for the detailed explanation @jgiardino . Based on this, is this functionality you would still want to have for toast alerts @mceledonia and @gdoyle1 ? It seems like it could work but also raises a11y complications.

**Comment by mcarrano** [comment]
@mceledonia I've assigned this to you for further consideration. Seems like there are some accessibility considerations and we should decide if this is something we want to get on the PatternFly roadmap or if there are alternate approaches.

**Comment by mceledonia** [comment]
Thanks again for the assessment @jgiardino! That's very helpful.

I'm personally a fan of option 2 since the user would be taught that they can find more details in the notification drawer after a single experience. If it is difficult at all for them to navigate to the notification drawer, that would be a separate issue that needs to be addressed. We could also tie the functionality of expandable toast alerts to the requirement of also having a drawer.

And it sounds like that doesn't apply to inline alert expansion since those are persistent, so I think we're all set there!

@mcarrano @gdoyle1 @jessiehuff

**Comment by mceledonia** [comment]
I do think it could be nice to have some additional functionality for toast alerts specifically for keyboard users. If we can detect navigation method (maybe a focus state triggers this?) we could pause the timeout on the toast to allow it to persist as long as needed until the user navigates away.

**Comment by gdoyle1** [comment]
@mceledonia For OpenShift, I don't think the connection with notification drawer would be possible. We only include warning and critical alerts in there as well as update recommendations today. We'd most likely only use toast notifications for states of actions related to events. For example, after the deletion of resources has completed, we'd use a success toast with "20 resources deleted" or something like that... same thing for updates, once an update finishes, "your cluster has been updated to version x.xx.xxx". Those types of messages wouldn't live in the notification drawer. Although, I could see us potentially having them live/captured in "Events"?

**Comment by mcarrano** [comment]
This is potentially needed for OS 4.9 (July 2 feature).

**Comment by gdoyle1** [comment]
I've come up with some proposals in [this Marvel](https://marvelapp.com/prototype/g8idd54)! @mcarrano @mceledonia @jgiardino @jessiehuff if you all could take a look - I added some specific items about accessibility in there too (and hopefully my wording was right 😊)

**Comment by mcarrano** [comment]
Looks like a good start @gdoyle1 . I do have some comments/questions. Is this really an alert or more like a modeless window that can float on top on the page? I ask because it seems like the use case is more to report status of ongoing tasks rather than momentary events. Is that right?

If it's something that will be kept open on the user's screen, than I wonder if it should be movable and/or have the ability to minimize without closing. I also wonder about the proper default location. Do you expect this would appear in the same location as a standard toast alert or does that need to be configurable? Let me know your thoughts.

**Comment by gdoyle1** [comment]
@mcarrano Yes it is more like a modeless window that can float on top of the page. And yes, it would be used to report ongoing tasks rather than momentary events! I agree it could be useful to move it around. You can collapse it currently although it won't be removed from the screen... I can explore minimizing it out of view as well.

Originally I was thinking it could appear in the same place as a standard toast alert but I felt like it being in the bottom right by default made more sense since it seemed less "in the way" than the top right. If it can be moved though... not sure if the default placement really matters at that point. I could see it being in the top right and have additional toast alerts stack on top of it (although that may be disruptive to have it in addition to the toasts)

**Comment by gdoyle1** [comment]
@mcarrano @mceledonia @jessiehuff I made some updates based on our conversation today! https://marvelapp.com/prototype/g8idd54 I think the inline alert is in a good spot and I spoke with other OpenShift designers who still think that it'd be useful as a variant. Would appreciate feedback on integrating the ongoing tasks list into the notification drawer!

**Comment by mcarrano** [comment]
@gdoyle1 I like the direction this is going. I agree that the expandable inline alert is in good shape and should be an easy enhancement. I left a couple of comments regarding the tasks in the drawer, but in general I like the direction. I think the tabs work to switch between Alerts and Tasks.

**Comment by gdoyle1** [comment]
@mcarrano I've finalized the wireframes for the task tracker here: https://marvelapp.com/prototype/g8idd54. It includes creating a new icon so I've created a pf-design issue for that in #1059 . Otherwise, this should be good to go!

**Comment by mcarrano** [comment]
Potentially consider this issue together with the task drawer modifications: https://github.com/patternfly/patternfly/issues/2993

**Comment by gdoyle1** [comment]
[This Marvel](https://marvelapp.com/prototype/g8idd54) is now up-to-date with the new masthead icon and the variations of how it should work. I added an example flow of how this could look as well! This will be dependent on #1059 being completed first, and I will be working on that this week

**Comment by gdoyle1** [comment]
I completed #1059 and created a PF core issue to get that icon into the webfont. I believe this issue is now ready to be moved to development @mcarrano !

**Comment by mcarrano** [comment]
I have opened the following core issues to begin implementation work:

https://github.com/patternfly/patternfly/issues/4186 to add tasks to the notification drawer component.
https://github.com/patternfly/patternfly/issues/4185 to add tasks variant to the notification badge.

**Comment by mcarrano** [comment]
@gdoyle1 can you paste a link to your Sketch artwork here so we can make that into a template? Thanks!

cc @doruskova

**Comment by gdoyle1** [comment]
Yes, of course! In [this folder ](https://drive.google.com/drive/folders/1rX-JM-fsxFpBnRmnr9RWFRCHyCyyOCWk?usp=sharing) you'll find the Sketch file for both task tracker and expandable inline alerts. I also included the svg for the new task icon which has already been added to PF core! @mcarrano @doruskova
